error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/StreamRules.java

```diff
@@ -250,6 +256,99 @@ Optional<S> after(Stream<T> stream, Function<? super T, S> function) {
     }
   }
 
+  // XXX: This rule assumes that any matched `Collector` does not perform any filtering.
+  static final class StreamCollectingAndThenCollectionIsEmpty<T> {
+    @BeforeTemplate
+    boolean before(Stream<T> stream, Collector<? super T, ?, ? extends Collection<?>> collector) {
+      return stream.collect(collectingAndThen(collector, Collection::isEmpty));
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
+    }
+  }
+
+  // XXX: This rule assumes that any matched `Collector` does not perform any filtering.
+  static final class StreamCollectingAndThenImmutableListIsEmpty<T> {
+    @BeforeTemplate
+    boolean before(
+        Stream<T> stream, Collector<? super T, ?, ? extends ImmutableList<?>> collector) {
+      return stream.collect(collectingAndThen(collector, ImmutableList::isEmpty));
+    }
+
+    @AfterTemplate
+    boolean after(Stream<T> stream) {
+      return stream.findAny().isEmpty();
+    }
+  }
```

Comment:
For each of these cases: the `::isEmpty` method reference may refer to a supertype of the type to which the the values are collected. That makes the number of possible combinations much larger. This relates to your second question; I'll propose something involving additional type parameters.

Reply:
Indeed, I really didn't like that solution. It looks much better now, thanks for the suggestion! Looks like I've got to do some homework on Generics 😄